File: src/main/java/io/wordii/nlp/NlpParser.java
package io.wordii.nlp;

import com.google.common.graph.ImmutableValueGraph;
import com.google.common.graph.MutableValueGraph;
import com.google.common.graph.ValueGraph;
import com.google.common.graph.ValueGraphBuilder;
import edu.stanford.nlp.coref.data.CorefChain;
import edu.stanford.nlp.ling.CoreLabel;
import edu.stanford.nlp.ling.IndexedWord;
import edu.stanford.nlp.pipeline.CoreDocument;
import edu.stanford.nlp.pipeline.CoreSentence;
import edu.stanford.nlp.pipeline.StanfordCoreNLP;
import edu.stanford.nlp.semgraph.SemanticGraph;
import edu.stanford.nlp.semgraph.SemanticGraphEdge;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

public final class NlpParser {
    private final StanfordCoreNLP pipeline;

    private NlpParser() {
        Properties props = new Properties();
        props.setProperty("annotators", "tokenize,ssplit,pos,lemma,ner,parse,coref");
        props.setProperty("coref.algorithm", "neural");
        pipeline = new StanfordCoreNLP(props);
    }

    public static NlpParser getInstance() {
        return Holder.INSTANCE;
    }

    private static final class Holder {
        private static final NlpParser INSTANCE = new NlpParser();
    }

    public SentenceSequence parse(String text) {
        CoreDocument coreDocument = new CoreDocument(text);
        pipeline.annotate(coreDocument);

        List<CoreSentence> sentences = coreDocument.sentences();
        List<Sentence> sentenceList = new ArrayList<>(sentences.size());
        for (int i = 0; i < sentences.size(); i++) {
            CoreSentence coreSentence = sentences.get(i);
            SemanticGraph semanticGraph = coreSentence.dependencyParse();
            ValueGraph<Word, Relation> valueGraph = mapToValueGraph(semanticGraph);
            List<Word> roots = semanticGraph.getRoots().stream()
                    .map(NlpParser::createWord)
                    .collect(Collectors.toUnmodifiableList());
            List<Word> wordList = coreSentence.tokens().stream()
                    .map(NlpParser::createWord)
                    .collect(Collectors.toUnmodifiableList());
            Sentence sentence = new Sentence(coreSentence.text(), valueGraph, roots, wordList, i + 1);
            sentenceList.add(sentence);
        }
        Set<CoreferenceChain> coreferenceChains = coreDocument.corefChains().values().stream()
                .map(corefChain -> new CoreferenceChain(createMention(sentenceList, corefChain.getRepresentativeMention()),
                        corefChain.getMentionsInTextualOrder().stream()
                                .map(mention -> createMention(sentenceList, mention))
                                .collect(Collectors.toUnmodifiableList())))
                .collect(Collectors.toUnmodifiableSet());
        return new SentenceSequence(Collections.unmodifiableList(sentenceList), coreferenceChains);
    }

    @NotNull
    private static ValueGraph<Word, Relation> mapToValueGraph(SemanticGraph semanticGraph) {
        MutableValueGraph<Word, Relation> valueGraph = ValueGraphBuilder
                .directed()
                .build();
        for (IndexedWord root : semanticGraph.getRoots()) {
            addChildNode(root, semanticGraph, valueGraph);
        }
        return ImmutableValueGraph.copyOf(valueGraph);
    }

    private static void addChildNode(IndexedWord root, SemanticGraph semanticGraph, MutableValueGraph<Word, Relation> valueGraph) {
        Word rootCopy = createWord(root);
        valueGraph.addNode(rootCopy);
        if (semanticGraph.hasChildren(root)) {
            Set<IndexedWord> children = semanticGraph.getChildren(root);
            for (IndexedWord child : children) {
                SemanticGraphEdge edge = semanticGraph.getEdge(root, child);
                Relation relation = Relation.valueOf(edge.getRelation().getShortName().toUpperCase());
                valueGraph.putEdgeValue(rootCopy, createWord(child), relation);
                addChildNode(child, semanticGraph, valueGraph);
            }
        }
    }

    @NotNull
    private static Word createWord(IndexedWord indexedWord) {
        return createWord(indexedWord.backingLabel());
    }

    @NotNull
    private static Word createWord(CoreLabel coreLabel) {
        return new Word(coreLabel.originalText(), coreLabel.lemma(), coreLabel.tag(), coreLabel.index());
    }

    @NotNull
    private static CoreferenceMention createMention(List<Sentence> sentenceList, CorefChain.CorefMention mention) {
        Word headWord = sentenceList.get(mention.sentNum - 1)
                .getWordList().get(mention.headIndex - 1); // headIndex - head word of the mention
        return new CoreferenceMention(mention.mentionSpan, headWord, new WordPosition(mention.sentNum, mention.headIndex));
    }
}

File: client-lib/src/main/kotlin/io/wordii/nlp/client/NlpParser.kt
package io.wordii.nlp.client

import com.google.common.graph.ImmutableValueGraph
import com.google.common.graph.MutableValueGraph
import com.google.common.graph.ValueGraph
import com.google.common.graph.ValueGraphBuilder
import edu.stanford.nlp.coref.data.CorefChain
import edu.stanford.nlp.ling.CoreLabel
import edu.stanford.nlp.ling.IndexedWord
import edu.stanford.nlp.pipeline.CoreDocument
import edu.stanford.nlp.pipeline.StanfordCoreNLP
import edu.stanford.nlp.semgraph.SemanticGraph
import io.wordii.nlp.api.*
import kotlinx.collections.immutable.toImmutableList
import kotlinx.collections.immutable.toImmutableSet
import java.util.*

object NlpParser {
    private val pipeline: StanfordCoreNLP by lazy {
        val props = Properties()
        props.setProperty("annotators", "tokenize,ssplit,pos,lemma,ner,parse,coref")
        props.setProperty("coref.algorithm", "neural")
        StanfordCoreNLP(props)
    }

    fun parse(text: String?): SentenceSequence {
        if (text == null) {
            return SentenceSequence(emptyList(), emptySet())
        }
        val coreDocument = CoreDocument(text)
        pipeline.annotate(coreDocument)
        val sentences = coreDocument.sentences()
        val sentenceList = mutableListOf<Sentence>()
        for ((i, coreSentence) in sentences.withIndex()) {
            val semanticGraph = coreSentence.dependencyParse()
            val valueGraph = mapToValueGraph(semanticGraph)
            val roots = semanticGraph.roots
                .map { indexedWord: IndexedWord -> createWord(indexedWord) }
                .toImmutableList()
            val wordList = coreSentence.tokens()
                .map { coreLabel: CoreLabel -> createWord(coreLabel) }
                .toImmutableList()
            val sentence = Sentence(coreSentence.text(), valueGraph, roots, wordList, i + 1)
            sentenceList.add(sentence)
        }
        val coreferenceChains = coreDocument.corefChains()
            .map { (_, corefChain: CorefChain) ->
                CoreferenceChain(createMention(sentenceList, corefChain.representativeMention),
                    corefChain.mentionsInTextualOrder
                        .map { mention: CorefChain.CorefMention -> createMention(sentenceList, mention) }
                        .toImmutableList())
            }
            .toImmutableSet()
        return SentenceSequence(sentenceList.toImmutableList(), coreferenceChains)
    }

    private fun mapToValueGraph(semanticGraph: SemanticGraph): ValueGraph<Word, Relation> {
        val valueGraph = ValueGraphBuilder
            .directed()
            .build<Word, Relation>()
        for (root in semanticGraph.roots) {
            addChildNode(root, semanticGraph, valueGraph)
        }
        return ImmutableValueGraph.copyOf(valueGraph)
    }

    private fun addChildNode(
        root: IndexedWord,
        semanticGraph: SemanticGraph,
        valueGraph: MutableValueGraph<Word, Relation>
    ) {
        val rootCopy = createWord(root)
        valueGraph.addNode(rootCopy)
        if (semanticGraph.hasChildren(root)) {
            val children = semanticGraph.getChildren(root)
            for (child in children) {
                val edge = semanticGraph.getEdge(root, child)
                val relation = Relation.valueOf(edge.relation.shortName.toUpperCase())
                valueGraph.putEdgeValue(rootCopy, createWord(child), relation)
                addChildNode(child, semanticGraph, valueGraph)
            }
        }
    }

    private fun createWord(indexedWord: IndexedWord): Word {
        return createWord(indexedWord.backingLabel())
    }

    private fun createWord(coreLabel: CoreLabel): Word {
        return Word(coreLabel.originalText(), coreLabel.lemma(), coreLabel.tag(), coreLabel.index())
    }

    private fun createMention(sentenceList: List<Sentence>, mention: CorefChain.CorefMention): CoreferenceMention {
        val headWord = sentenceList[mention.sentNum - 1]
            .wordList[mention.headIndex - 1] // headIndex - head word of the mention
        return CoreferenceMention(mention.mentionSpan, headWord, WordPosition(mention.sentNum, mention.headIndex))
    }
}
